File: service/src/tasks/ingest.py
from sqlalchemy import exc
from models.database import Session
from models import FeedItem
import feedparser
from tqdm import tqdm
from datetime import datetime


def main():
    # Parse the RSS feed
    print("Fetching RSS feed")
    d = feedparser.parse("https://nitter.net/i/lists/1660595980486492162/rss")

    print(f"Processing {len(d.entries)} items")
    # Open a new session
    session = Session()

    inserted_count = 0
    skipped_count = 0

    # Loop over each entry
    for entry in d.entries:
        date_format = "%a, %d %b %Y %H:%M:%S %Z"
        parsed_time = datetime.strptime(
            entry.published, date_format)
        feed_item = FeedItem(
            title_raw=entry.title,
            link=entry.link,
            description_raw=entry.description,
            published=parsed_time,
            author=entry.author
        )

        session.add(feed_item)
        try:
            session.commit()
            inserted_count += 1
        except exc.IntegrityError:
            skipped_count += 1
            session.rollback()

    print(f"Inserted {inserted_count} items, skipped {skipped_count} items")

    # Close the session
    session.close()


if __name__ == "__main__":
    main()

File: app/jobs/ingest.ts
import { PrismaClient } from "@prisma/client"
import dayjs from "dayjs"
import Parser from "rss-parser"

const prisma = new PrismaClient()

async function main() {
  console.log("Fetching RSS feed")
  const parser = new Parser()
  const feed = await parser.parseURL("https://nitter.net/i/lists/1660595980486492162/rss")

  console.log(`Processing ${feed.items.length} items`)

  let insertedCount = 0
  let skippedCount = 0

  for (const item of feed.items) {
    const parsedTime = dayjs(item.pubdate).toDate()

    try {
      await prisma.feed_items.create({
        data: {
          title_raw: item.title!,
          link: item.link!,
          description_raw: item.content!,
          published: parsedTime,
          author: item.creator!,
        },
      })
      insertedCount += 1
    } catch (error) {
      // console.error(error)
      skippedCount += 1
    }
  }

  console.log(`Inserted ${insertedCount} items, skipped ${skippedCount} items`)
}

main()
  .catch((e) => {
    throw e
  })
  .finally(async () => {
    await prisma.$disconnect()
  })
